### PR TITLE
ENT-5850: Modified classesmatching() function to search parent bundles with inherit => true

### PR DESCRIPTION
**Pull request by craigcomstock**

### CODE REVIEW

**Comment by craigcomstock** [comment]
@cf-bottom jenkins with exotics, why not!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9636/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9636/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9636/

**Comment by craigcomstock** [comment]
I'll check the two failures: in windows and is_readable.

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 1]
```diff
@@ -0,0 +1,38 @@
+bundle agent __main__
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:1]: Unindent everything after this line, once

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  {
+     classes:
+
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:4]: Excessive newline character

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      "defined_in_parent" expression => "cfengine";
+
+     methods:
+      "ENT-5850"
+        usebundle => ENT_5850,
+        inherit => "true";
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:10]:
```suggestion
     methods:
       "ENT-5850"
         usebundle => ENT_5850,
         inherit => "true";
```
Inconsistent indent size

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  bundle agent  ENT_5850
+  {
+    vars:
+        "c_matching"
+          slist => classesmatching(".*");
+        "c_matching_defined_in_parent"
+          slist => classesmatching("defined_in_parent");
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:18]:
```suggestion
    vars:
      "c_matching"
        slist => classesmatching(".*");
      "c_matching_defined_in_parent"
        slist => classesmatching("defined_in_parent");
```

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        "$(this.promise_filename) Pass"
+          if => some( "defined_in_parent", c_matching);
+
+        "$(this.promise_filename) FAIL $(this.promise_filename)$(const.n)Could not find class 'defined_in_parent' in classesmatching() but it's defined"
+          if => and( not( some( "defined_in_parent", c_matching) ),
+                     defined_in_parent
+          );
+
+       "Classes found by classesmatching() starting with 'defined' $(with)"
+        with => join( ", ", classesmatching("defined.*") );
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:33]:
```suggestion
         with => join( ", ", classesmatching("defined.*") );
```

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+     classes:
+
+      "defined_in_parent" expression => "cfengine";
+
+     methods:
+      "ENT-5850"
+        usebundle => ENT_5850,
+        inherit => "true";
+  }
+  bundle agent  ENT_5850
```
Comment (larsewi) [inline tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:12]:
```suggestion
  }
 
  bundle agent ENT_5850
```

File: libpromises/eval_context.c [hunk 7]
```diff
@@ -3484,6 +3484,47 @@ bool EvalContextIsIgnoringLocks(const EvalContext *ctx)
     return ctx->ignore_locks;
 }
 
+StringSet *ClassesMatchingLocalRecursive(const EvalContext *ctx, const char* regex, const Rlist *tags, bool first_only, size_t stack_index)
+{
+  assert(ctx != NULL);
```
Comment (larsewi) [inline libpromises/eval_context.c:3489]: Indents should be four spaces

File: libpromises/eval_context.c [hunk 8]
```diff
@@ -3484,6 +3484,47 @@ bool EvalContextIsIgnoringLocks(const EvalContext *ctx)
...
+  }
+
+  StringSet *parent_matches = StringSetNew();
+  if (stack_index > 0 && frame->inherits_previous)
+  {
+    StringSetDestroy(parent_matches);
+    parent_matches = ClassesMatchingLocalRecursive(ctx, regex, tags, first_only, stack_index - 1);
+  }
+  StringSetJoin(matches, parent_matches, xstrdup);
+  StringSetDestroy(parent_matches);
```
Comment (larsewi) [inline libpromises/eval_context.c:3510]:
```suggestion
  if (stack_index > 0 && frame->inherits_previous)
  {
    StringSet *parent_matches = ClassesMatchingLocalRecursive(ctx, regex, tags, first_only, stack_index - 1);
    StringSetJoin(matches, parent_matches, xstrdup);
    StringSetDestroy(parent_matches);
  }

```
The `parent_matches` variable can live within the scope of the if statement

**Review by larsewi — CHANGES_REQUESTED** [review]
Nice work 🚀  Only some nitpicks

File: libpromises/eval_context.c (hunk 7 above)
Comment (craigcomstock) [reply libpromises/eval_context.c:3489]: I know! I tried running clang-format yesterday and it mangled the entire file horribly. So was about to just clang-format my changes and then got distracted. I am a bit surprised we don't have a CI tasks/github action for checking style :(

File: libpromises/eval_context.c (hunk 8 above)
Comment (craigcomstock) [reply libpromises/eval_context.c:3510]: ah, right, good one.

File: tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf (hunk 1 above)
Comment (craigcomstock) [reply tests/acceptance/02_classes/02_functions/classesmatching_inherit.cf:1]: haha! says the guy who worked on a cfengine policy language parser/formatter. ;)

**Review by nickanderson — DISMISSED** [review]
Tests look ok to me,

**Review by larsewi — DISMISSED** [review]
LGTM 🚀 Make sure it passes Jenkins

**Comment by craigcomstock** [comment]
> LGTM 🚀 Make sure it passes Jenkins

It did already once with two failures: windows in github and is_readable in jenkins on some exotic.

I will manually debug on windows tomorrow and then re-submit to github/jenkins.

**Comment by craigcomstock** [comment]
the failure on windows is due to the fact that the github action tests against the most recent master enterprise binaries instead of against the code changes in this PR, so can be ignored since it is a test I am introducing which is failing.

isreadable passed for me on a linux box so known issue on hpux: https://northerntech.atlassian.net/browse/ENT-9930

**Review by nickanderson — APPROVED** [review]

**Comment by craigcomstock** [comment]
cherry picks to
3.18.x https://github.com/cfengine/core/pull/5298
3.21.x https://github.com/cfengine/core/pull/5300

**Pull request merged**